src/app/qgisapp.cpp

```diff
@@ -6500,7 +6500,42 @@ void QgisApp::newMeshLayer()
 void QgisApp::showRasterCalculator()
 {
   QgsRasterCalcDialog d( qobject_cast<QgsRasterLayer *>( activeLayer() ), this );
-  if ( d.exec() == QDialog::Accepted )
+  if ( d.exec() != QDialog::Accepted )
+  {
+    return;
+  }
+  if ( d.useVirtualProvider() )
+  {
+    QgsRasterDataProvider::VirtualRasterParameters virtualCalcParams;
+    virtualCalcParams.crs = d.outputCrs();
+    virtualCalcParams.extent = d.outputRectangle();
+    virtualCalcParams.width = d.numberOfColumns();
+    virtualCalcParams.height = d.numberOfRows();
+    virtualCalcParams.formula = d.formulaString();
+
+    QStringList rLayerDictionary = QgsRasterCalcNode::referencedLayerNames( d.formulaString() );
+    QSet<QString> uniqueRasterUriTmp;
+
+    for ( const auto &r : QgsRasterCalculatorEntry::rasterEntries() )
+    {
+
+      if ( ( ! rLayerDictionary.contains( r.raster->name() ) ) || uniqueRasterUriTmp.contains( r.raster->publicSource() ) ) continue;
+      uniqueRasterUriTmp.insert( r.raster->publicSource() );
+
+      QgsRasterDataProvider::VirtualRasterInputLayers projectRLayer;
+      projectRLayer.name = r.raster->name();
+      projectRLayer.provider = r.raster->dataProvider()->name();
+      projectRLayer.uri = r.raster->publicSource();
+
+      virtualCalcParams.rInputLayers.append( projectRLayer );
+    }
+
+    addRasterLayer( QgsRasterDataProvider::encodeVirtualRasterProviderUri( virtualCalcParams ),
+                    //QStringLiteral( "virtuallayer" ),
+                    d.virtualLayerName() == QString( "" ) ? d.formulaString() : d.virtualLayerName(),
```

Comment:
rather than `str == QString( "" )` it is better to do `str.isEmpty()` as this removes creation of a temporary QString